File: src/main/rs/emulate/legacy/config/ConfigEncoder.java
package rs.emulate.legacy.config;

import java.io.ByteArrayOutputStream;
import java.util.List;
import java.util.Map;

import rs.emulate.legacy.archive.Archive;
import rs.emulate.legacy.archive.ArchiveEntry;
import rs.emulate.legacy.archive.ArchiveUtils;
import rs.emulate.shared.prop.DefinitionProperty;
import rs.emulate.shared.util.DataBuffer;

/**
 * Encodes data from {@link ArchiveEntry} objects in the {@code config} {@link Archive}.
 *
 * @author Major
 *
 * @param <T> The type the data is encoded from.
 */
public final class ConfigEncoder<T extends MutableConfigDefinition> {

	/**
	 * Returns whether or not the specified {@link Map} entry contains a {@link DefinitionProperty} that should be
	 * encoded.
	 * 
	 * @param entry The Map entry.
	 * @return {@code true} if the property should be encoded, {@code false} if not.
	 */
	private static boolean validProperty(Map.Entry<Integer, DefinitionProperty<?>> entry) {
		return entry.getValue().valuePresent();
	}

	/**
	 * The List of definitions to encode.
	 */
	protected final List<T> definitions;

	/**
	 * The name of the ArchiveEntry.
	 */
	protected final String name;

	/**
	 * Creates the ConfigEncoder.
	 *
	 * @param name The name of the {@link ArchiveEntry}.
	 * @param definitions The {@link List} of definitions to encode.
	 */
	public ConfigEncoder(String name, List<T> definitions) {
		this.name = name;
		this.definitions = definitions;
	}

	/**
	 * Encodes the {@link List} of {@code T}s into an {@link ArchiveEntry}, and returns a new {@link Archive} containing
	 * the entries in the specified Archive, and the newly created one.
	 *
	 * @param archive The Archive to use as a base.
	 * @return The new Archive.
	 */
	public Archive encodeInto(Archive archive) {
		return archive.addEntries(encode());
	}

	/**
	 * Encodes the {@link List} of {@code T}s into an {@link ArchiveEntry}.
	 *
	 * @return The archive entry.
	 */
	private ArchiveEntry[] encode() {
		ByteArrayOutputStream data = new ByteArrayOutputStream();
		ByteArrayOutputStream index = new ByteArrayOutputStream();

		int size = definitions.size(), last = 2;
		writeShort(data, size);
		writeShort(index, size);

		for (T definition : definitions) {
			definition.getProperties().stream().filter(ConfigEncoder::validProperty).forEach(entry -> write(entry, data));
			data.write(ConfigConstants.DEFINITION_TERMINATOR);

			int change = data.size() - last;
			writeShort(index, change);
			last = data.size();
		}

		int hash = ArchiveUtils.hash(name + ConfigConstants.DATA_EXTENSION);
		ArchiveEntry dataEntry = new ArchiveEntry(hash, DataBuffer.wrap(data.toByteArray()));

		hash = ArchiveUtils.hash(name + ConfigConstants.INDEX_EXTENSION);
		ArchiveEntry indexEntry = new ArchiveEntry(hash, DataBuffer.wrap(index.toByteArray()));

		return new ArchiveEntry[] { dataEntry, indexEntry };
	}

	/**
	 * Writes the specified {@link Map} entry containing the opcode and {@link DefinitionProperty} to the specified
	 * {@link ByteArrayOutputStream}.
	 *
	 * @param entry The map entry.
	 * @param os The byte array output stream.
	 */
	private void write(Map.Entry<Integer, DefinitionProperty<?>> entry, ByteArrayOutputStream os) {
		DataBuffer buffer = entry.getValue().encode();
		byte[] bytes = buffer.getBytes();

		os.write(entry.getKey());
		os.write(bytes, 0, bytes.length); // write(byte[]) throws IOException unnecessarily
	}

	/**
	 * Writes a {@code short} (i.e. a 16-bit value) to the specified {@link ByteArrayOutputStream}.
	 * 
	 * @param stream The ByteArrayOutputStream.
	 * @param value The value to write.
	 */
	private void writeShort(ByteArrayOutputStream stream, int value) { // TODO move to utils class
		byte[] bytes = { (byte) (value >> 8), (byte) value };
		stream.write(bytes, 0, bytes.length); // write(byte[]) throws IOException unnecessarily
	}

}

File: legacy/src/main/rs/emulate/legacy/config/ConfigEncoder.kt
package rs.emulate.legacy.config

import rs.emulate.legacy.archive.Archive
import rs.emulate.legacy.archive.ArchiveEntry
import rs.emulate.legacy.archive.ArchiveUtils
import rs.emulate.shared.util.DataBuffer

import java.io.ByteArrayOutputStream
import kotlin.collections.Map.Entry

/**
 * Encodes data from [ArchiveEntry] objects in the `config` [Archive].
 *
 * @param name The name of the [ArchiveEntry].
 * @param definitions The [List] of definitions to encode.
 * @param T The type the data is encoded from.
 */
class ConfigEncoder<out T : MutableConfigDefinition>(private val name: String, private val definitions: List<T>) {

    /**
     * Encodes the [List] of `T`s into an [ArchiveEntry].
     *
     * @return The archive entry.
     */
    fun encode(): Array<ArchiveEntry> {
        val data = ByteArrayOutputStream()
        val index = ByteArrayOutputStream()

        val size = definitions.size
        var last = 2
        writeShort(data, size)
        writeShort(index, size)

        for (definition in definitions) {
            definition.serializableProperties()
                .filter(this::validProperty)
                .forEach { write(it, data) }
            data.write(ConfigConstants.DEFINITION_TERMINATOR)

            val change = data.size() - last
            writeShort(index, change)
            last = data.size()
        }

        var hash = ArchiveUtils.hash(name + ConfigConstants.DATA_EXTENSION)
        val dataEntry = ArchiveEntry(hash, DataBuffer.wrap(data.toByteArray()))

        hash = ArchiveUtils.hash(name + ConfigConstants.INDEX_EXTENSION)
        val indexEntry = ArchiveEntry(hash, DataBuffer.wrap(index.toByteArray()))

        return arrayOf(dataEntry, indexEntry)
    }

    /**
     * Encodes the [List] of `T`s into an [ArchiveEntry], and returns a new [Archive] containing
     * the entries in the specified Archive, and the newly created one.
     */
    fun encodeInto(archive: Archive): Archive {
        return archive.addEntries(*encode())
    }

    /**
     * Returns whether or not the specified [Map] entry contains a [SerializableProperty] that should be
     * encoded.
     */
    private fun validProperty(entry: Entry<Int, SerializableProperty<*>>): Boolean {
        return entry.value.valuePresent()
    }

    /**
     * Writes the specified [Map] entry containing the opcode and [SerializableProperty] to the specified
     * [ByteArrayOutputStream].
     */
    private fun write(entry: Entry<Int, SerializableProperty<*>>, os: ByteArrayOutputStream) {
        val buffer = entry.value.encode()
        val bytes = buffer.remainingBytes

        os.write(entry.key)
        os.write(bytes)
    }

    /**
     * Writes a `short` (i.e. a 16-bit value) to the specified [ByteArrayOutputStream].
     */
    private fun writeShort(stream: ByteArrayOutputStream, value: Int) {
        stream.write(value shr 8)
        stream.write(value)
    }

}
